### PR TITLE
adding support for private ip + zones

### PR DESCRIPTION
**Pull request by gene-redpanda**
Before going into the features, I made an important change to the base module that should be considered carefully:

1) SGs are now coalesce rather than concat:

I did this because when attempting to create a private/public split with the original module's concat, the client node was being added to both the private and public SG. This caused numerous issues that disappeared with the change to coalesce. It is still possible to replicate the original functionality by creating your own subnets and specifying them in a list but it isn't a default feature. 

2) subnets are now specified per instance type rather than a single subnet variable

It wasn't possible to specify different subnets for different instance types which made it impossible to specify a public jumpbox + private cluster. As an additional feature, for instances outside of the redpanda broker group it will attempt to use the subnet variable specific to the instance, then the subnet variable specific to the redpanda broker group, then the default subnet. 

*** 

Adds support for two more models as well as examples demoing them that more perfectly replicate likely client network configurations as part of supporting ansible side work on airgap.

The two new models are:

* private brokers without net access accessible to public access jumpbox while using the default vpc still
* same as above, but with created vpc instead of defaults, r53 records, no public ips on the broker

My hope is that this is both more useful for clients -- because these are configs end users are likely to use -- and improves our ability to test in a realistic manner.

As a final note, I tried to include a default sg ingress/egress ruleset that was merged with any user supplied sg ingress/egress rules unless a bool was flipped to specify not to do the merge, but after substantial testing wasn't able to come up with a method that wasn't rejected out of hand by terraform due to type issues and eventually scrapped the idea due to time constraints. I'll try again at a later date since this would be convenient and I'm reasonably sure there's a way to get it done.

### CODE REVIEW

**Comment by gene-redpanda** [comment]
I've tested from init to apply and then pinging and sshing around to the various instances from the jumpbox locally, but haven't configured the buildkite pipeline yet. Please ignore the CI failure.

File: examples/vpc/main.tf [hunk 1]
```diff
@@ -41,36 +50,181 @@ resource "aws_route_table" "test" {
...
+  route_table_id = aws_route_table.test.id
 }
 
+resource "aws_security_group" "client_sec_group" {
+  name        = "${var.deployment_prefix}-client-sg"
+  description = "client security group"
+  tags        = var.tags
+  vpc_id      = aws_vpc.test.id
+  ingress {
+    description = "Allow oubound to ssh"
```
Comment (r-vasquez) [inline examples/vpc/main.tf]:
```suggestion
    description = "Allow outbound to ssh"
```

File: examples/private/main.tf [hunk 2]
```diff
@@ -0,0 +1,224 @@
+resource "aws_security_group" "client_sec_group" {
+  name        = "${var.deployment_prefix}-client-sg"
+  description = "client security group"
+  vpc_id      = var.vpc_id
+  tags        = var.tags
+
+  ingress {
+    description = "Allow oubound to ssh"
```
Comment (r-vasquez) [inline examples/private/main.tf]:
```suggestion
    description = "Allow outbound to ssh"
```

File: examples/private/main.tf [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    from_port   = 22
+    to_port     = 22
+    protocol    = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+
+  egress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = "-1"
```
Comment (r-vasquez) [inline examples/private/main.tf:18]: From https://registry.terraform.io/providers/hashicorp/aws/3.42.0/docs/resources/security_group_rule

> Setting protocol = "all" or protocol = -1 with from_port and to_port will result in the EC2 API creating a security group rule with all ports open. This API behavior cannot be controlled by Terraform and may generate warnings in the future.

File: examples/private/main.tf [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  hosts_file                      = var.hosts_file
+  tags                            = var.tags
+  clients                         = 1
+  security_groups_client          = [aws_security_group.client_sec_group.id]
+  associate_public_ip_addr        = true
+  associate_public_ip_addr_client = true
+  availability_zone               = ["us-west-2a"]
+  egress_rules                    = {
+    "SSH" = {
+      description     = "Allow oubound to ssh"
```
Comment (r-vasquez) [inline examples/private/main.tf]:
```suggestion
      description     = "Allow outbound to ssh"
```

File: examples/private/main.tf [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    "RPC" = {
+      description     = "Allow security-group only to access Redpanda RPC endpoint for intra-cluster communication"
+      from_port       = 33145
+      to_port         = 33145
+      protocol        = "tcp"
+      enabled         = true
+      cidr_blocks     = null
+      self            = true
+      security_groups = [aws_security_group.client_sec_group.id]
+    }
```
Comment (r-vasquez) [inline examples/private/main.tf]: Question out of terraform ignorance:

You mention: `Allow anywhere inbound to access the Redpanda Kafka endpoint` and set the security groups, and say: `Allow security-group only to access Redpanda RPC endpoint for intra-cluster communication` with the same configuration.

Aren't both the same? aren't both `Allow security-group only to access Redpanda <Kafka Endpoint/RPC endpoint>`?

File: examples/private/main.tf [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+variable "nodes" {
+  type    = number
+  default = 3
+}
+
+variable "deployment_prefix" {
+  type    = string
+  default = "test-rp-cluster"
```
Comment (r-vasquez) [inline examples/private/main.tf]: Why test? Isn't this prod-ready usable for others? (I know they can change it, but just asking about the default)

File: examples/private/main.tf [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+variable "enable_monitoring" {
+  type    = bool
+  default = true
+}
```
Comment (r-vasquez) [inline examples/private/main.tf:93]: [Question] What does this change in the VM deployment?

File: examples/vpc/main.tf [hunk 8]
```diff
@@ -41,36 +50,181 @@ resource "aws_route_table" "test" {
...
-  records = [element(values(module.redpanda-cluster.redpanda_map), count.index)]
+module "redpanda-cluster" {
+  source                          = "../../"
+  public_key_path                 = var.public_key_path
+  nodes                           = var.nodes
+  deployment_prefix               = var.deployment_prefix
+  enable_monitoring               = var.enable_monitoring
+  tiered_storage_enabled          = var.tiered_storage_enabled
+  allow_force_destroy             = var.allow_force_destroy
+  vpc_id                          = aws_vpc.test.id
```
Comment (r-vasquez) [inline examples/vpc/main.tf]: Shouldn't this point to the variable?:

```
variable "vpc_id" {
  description = "only set when you are planning to provide your own network rather than using the default one"
  type        = string
  default     = ""
}
```

File: examples/private/main.tf (hunk 3 above)
Comment (gene-redpanda) [reply examples/private/main.tf:18]: Yes, that is intended for the client SG. I may harden it in future but I wanted it to be very accessible for testing. The brokers are all well protected.

File: examples/private/main.tf (hunk 5 above)
Comment (gene-redpanda) [reply examples/private/main.tf]: Ah shoot I forgot to update the descriptions. They're all copy pasted from the module so that I could associate them with the client SG, but some of them have been changed from all open to just within the server SG for hardening purposes.

File: examples/private/main.tf (hunk 6 above)
Comment (gene-redpanda) [reply examples/private/main.tf]: mostly so i could have a non-colliding static prefix. i can change it 🤷‍♀️ 

I would say this is productionish. I wouldn't recommend it as there's a lot more stuff you could do to make it more secure and robust, but you could get away with it if you needed to do so.

File: examples/private/main.tf (hunk 7 above)
Comment (gene-redpanda) [reply examples/private/main.tf:93]: Turns on the prometheus node. I wanted to check on it too.

File: examples/vpc/main.tf (hunk 8 above)
Comment (gene-redpanda) [reply examples/vpc/main.tf]: Not in this instance as we're actually creating the VPC in this one. The variable should probably be deleted.

File: .buildkite/pipeline.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      - docker#v5.4.0:
+          image: glrp/atgt:latest
+          environment:
+            - DA_AWS_ACCESS_KEY_ID
+            - DA_AWS_SECRET_ACCESS_KEY
+            - AWS_DEFAULT_REGION
+
+  - label: Terraform Apply and Destroy Simple
+    key: apply-destroy-simple
+    command: .buildkite/scripts/apply.sh --TF_DIR examples/simple --PREFIX rp-simple --AWS_ACCESS_KEY_ID $DA_AWS_ACCESS_KEY_ID --AWS_SECRET_ACCESS_KEY $DA_AWS_SECRET_ACCESS_KEY
```
Comment (gousteris) [inline .buildkite/pipeline.yaml]:
```suggestion
    command: .buildkite/scripts/apply.sh --TF_DIR examples/simple --PREFIX rp-simple --AWS_ACCESS_KEY_ID $$DA_AWS_ACCESS_KEY_ID --AWS_SECRET_ACCESS_KEY $$DA_AWS_SECRET_ACCESS_KEY
```

ie doubling the `$` character
https://buildkite.com/docs/pipelines/environment-variables#runtime-variable-interpolation

File: .buildkite/scripts/apply.sh [hunk 10]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+
+AWS_ACCESS_KEY_ID=$DA_AWS_ACCESS_KEY_ID
+AWS_SECRET_ACCESS_KEY=$DA_AWS_SECRET_ACCESS_KEY
```
Comment (gousteris) [inline .buildkite/scripts/apply.sh]:
```suggestion
export AWS_ACCESS_KEY_ID=$DA_AWS_ACCESS_KEY_ID
export AWS_SECRET_ACCESS_KEY=$DA_AWS_SECRET_ACCESS_KEY
```

**Comment by gene-redpanda** [comment]
Added a fix for https://github.com/redpanda-data/terraform-aws-redpanda-cluster/issues/6

File: examples/private/main.tf [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+  description = "client security group"
+  vpc_id      = var.vpc_id
+  tags        = var.tags
+
+  ingress {
+    description = "Allow outbound to ssh"
+    from_port   = 22
+    to_port     = 22
+    protocol    = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/private/main.tf:12]: I'd like to see us start controlling inbound/outbound access a little better. can we make ssh inbound access a variable that defaults to 0.0.0.0/0, rather than hard coding?

File: examples/private/main.tf [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+  }
+
+  egress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/private/main.tf:19]: anywhere we stick an open rule like this, we should consider making it a variable.

File: examples/private/main.tf [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+  }
+}
+module "redpanda-cluster" {
+  source                          = "../../"
```
Comment (hcoyote) [inline examples/private/main.tf:24]: This seems strange. what's it actually sourcing?

File: examples/private/main.tf [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+  associate_public_ip_addr_client = true
+  availability_zone               = ["us-west-2a"]
+  egress_rules                    = {
+    "SSH" = {
+      description     = "Allow outbound to ssh"
+      from_port       = 22
+      to_port         = 22
+      protocol        = "tcp"
+      enabled         = true
+      cidr_blocks     = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/private/main.tf:46]: see comment previously about 0.0.0.0

File: examples/private/main.tf [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+    }
+  }
+  ingress_rules = {
+    "SSH" = {
+      description     = "Allow inbound to ssh"
+      from_port       = 22
+      to_port         = 22
+      protocol        = "tcp"
+      enabled         = true
+      cidr_blocks     = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/private/main.tf:58]: see comment previously about 0.0.0.0

File: examples/private/main.tf [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    "JavaOMB" = {
+      description     = "Allow anywhere inbound to access for Open Messaging Benchmark"
+      from_port       = 8080
+      to_port         = 8080
+      protocol        = "tcp"
+      enabled         = true
+      cidr_blocks     = null
+      self            = true
+      security_groups = [aws_security_group.client_sec_group.id]
+    }
```
Comment (hcoyote) [inline examples/private/main.tf]: I think omb access should be optional.

File: examples/public-vpc/main.tf [hunk 17]
```diff
@@ -0,0 +1,129 @@
+resource "aws_vpc" "test" {
```
Comment (hcoyote) [inline examples/public-vpc/main.tf:1]: Can we rename this vpc identifier to something more descriptive than test?

File: examples/public-vpc/main.tf [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+
+variable "tiered_storage_enabled" {
+  type    = bool
+  default = false
+}
+
+variable "allow_force_destroy" {
+  type    = bool
+  default = false
+}
```
Comment (hcoyote) [inline examples/public-vpc/main.tf]: Can we add descriptions to every variable that's missing them? If we need to do that in a different PR, that's fine. But we should have these examples pretty well defined and self-documenting

File: examples/vpc/main.tf [hunk 19]
```diff
@@ -16,11 +16,20 @@ resource "aws_route53_zone" "test" {
...
 
 
-resource "aws_subnet" "test" {
+resource "aws_subnet" "server" {
   vpc_id     = aws_vpc.test.id
   cidr_block = "10.0.1.0/24"
 
-  tags = var.tags
+  tags              = var.tags
+  availability_zone = "us-west-2a"
```
Comment (hcoyote) [inline examples/vpc/main.tf:11]: should this be set via a variable instead?

File: examples/vpc/main.tf [hunk 20]
```diff
@@ -41,36 +50,182 @@ resource "aws_route_table" "test" {
...
+  name        = "${var.deployment_prefix}-client-sg"
+  description = "client security group"
+  tags        = var.tags
+  vpc_id      = aws_vpc.test.id
+  ingress {
+    description = "Allow outbound to ssh"
+    from_port   = 22
+    to_port     = 22
+    protocol    = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/vpc/main.tf]: see previous comments about 0.0.0.0

File: examples/vpc/main.tf [hunk 21]
```diff
@@ -41,36 +50,182 @@ resource "aws_route_table" "test" {
...
+  }
 
-resource "aws_route53_record" "private_record" {
-  count = var.nodes
+  egress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/vpc/main.tf]: see previous comments about 0.0.0.0

File: examples/vpc/main.tf [hunk 22]
```diff
@@ -41,36 +50,182 @@ resource "aws_route_table" "test" {
...
+  clients                         = 1
+  availability_zone               = ["us-west-2a"]
+  egress_rules                    = {
+    "SSH" = {
+      description     = "Allow oubound to ssh"
+      from_port       = 22
+      to_port         = 22
+      protocol        = "tcp"
+      enabled         = true
+      cidr_blocks     = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/vpc/main.tf]: see previous comments about 0.0.0.0

File: examples/vpc/main.tf [hunk 23]
```diff
@@ -41,36 +50,182 @@ resource "aws_route_table" "test" {
...
+      self            = null
+      security_groups = [aws_security_group.client_sec_group.id]
+    }
+    "PING" = {
+      description     = "Allow ICMP"
+      from_port       = -1
+      to_port         = -1
+      protocol        = "icmp"
+      enabled         = true
+      cidr_blocks     = ["0.0.0.0/0"]
```
Comment (hcoyote) [inline examples/vpc/main.tf]: see previous comments about 0.0.0.0

File: examples/vpc/main.tf [hunk 24]
```diff
@@ -41,36 +50,182 @@ resource "aws_route_table" "test" {
...
+    "JavaOMB" = {
+      description     = "Allow anywhere inbound to access for Open Messaging Benchmark"
+      from_port       = 8080
+      to_port         = 8080
+      protocol        = "tcp"
+      enabled         = true
+      cidr_blocks     = null
+      self            = true
+      security_groups = [aws_security_group.client_sec_group.id]
+    }
```
Comment (hcoyote) [inline examples/vpc/main.tf]: see previous about OMB access

File: outputs.tf [hunk 25]
```diff
@@ -22,16 +22,9 @@ resource "random_id" "redpanda" {
   }
 }
 
-output "redpanda_map" {
-  value = { for i in range(length(aws_instance.redpanda[*].id)) :
-    lower(random_id.redpanda[i].b64_url) => aws_instance.redpanda[i].private_ip
-  }
-  description = "A map of random IDs to public IPs for the Redpanda instances."
-}
-
```
Comment (hcoyote) [inline outputs.tf:31]: not sure I understand the removal of the outputs.

**Review by hcoyote — CHANGES_REQUESTED** [review]

File: examples/private/main.tf (hunk 11 above)
Comment (gene-redpanda) [reply examples/private/main.tf:12]: This is an example tho, not part of the module. it's also not hard code?

File: examples/private/main.tf (hunk 13 above)
Comment (gene-redpanda) [reply examples/private/main.tf:24]: it's sourcing the module itself so that the example can run the local code rather than pulling from tf registry

File: outputs.tf (hunk 25 above)
Comment (gene-redpanda) [reply outputs.tf:31]: it was something i was testing. zero value add so i went a different route

File: examples/public-vpc/main.tf (hunk 17 above)
Comment (gene-redpanda) [reply examples/public-vpc/main.tf:1]: I want to make sure no one is going direct to prod with our examples so naming everything test is part of how I'm achieving it. My hope is that it embarrasses people into at least making light edits to change the names 😅. We're also sourcing the module from the repo rather than terraform's module registry (which is best practice for examples that are dual purposed to CI) so that should hopefully put another barrier in the way of copypasta

**Review by r-vasquez — APPROVED** [review]
LGTM

File: examples/public-vpc/main.tf (hunk 17 above)
Comment (hcoyote) [reply examples/public-vpc/main.tf:1]: fair enough.

**Review by hcoyote — APPROVED** [review]

File: client.tf [hunk 26]
```diff
@@ -1,11 +1,11 @@
 resource "aws_instance" "client" {
-  count                       = var.clients
+  count                       = var.client_count  * length(var.availability_zone)
   ami                         = coalesce(var.client_ami, data.aws_ami.ami.image_id)
   instance_type               = var.client_instance_type
   key_name                    = aws_key_pair.ssh.key_name
-  subnet_id                   = var.subnet_id
-  vpc_security_group_ids      = concat([aws_security_group.node_sec_group.id], var.security_groups_client)
-  associate_public_ip_address = var.associate_public_ip_addr
+  subnet_id                   = try(lookup(local.merged_subnets["client"], var.availability_zone[count.index % length(var.availability_zone)]), "")
```
Comment (hcoyote) [inline client.tf:6]: Not quite sure I understand what's happening here.

We index of the client instance and mod it by the number of AZs to come up with the name of an AZ, and then look that up in the map of merged subnets to figure out what subnet id to use for that az?

File: examples/private-vpc-no-zone/README.md [hunk 27]
```diff
@@ -0,0 +1,7 @@
+This example shows a best practices deployment in which the brokers are isolated in a separate subnet on a non default
+VPC and not allocated public addresses with any external traffic flowing through the internet gateway. The client can be
+configured as a proxy to allow access between users and the brokers, or additional security groups can be added to each
+port in the ingress/egress rulesets.
+
+This does not include a zone or address assignment to hosts, so may not be suitable for production as hosts will have to
```
Comment (hcoyote) [inline examples/private-vpc-no-zone/README.md:6]: by zone you mean dns zone for host->ip mapping?

File: examples/private-vpc-no-zone/main.tf [hunk 28]
```diff
@@ -0,0 +1,299 @@
...
+    "Grafana" = {
+      description     = "Allow anywhere inbound to access grafana end point for monitoring"
+      from_port       = 3000
+      to_port         = 3000
+      protocol        = "tcp"
+      enabled         = true
+      cidr_blocks     = null
+      self            = true
+      security_groups = [aws_security_group.client_sec_group.id]
+    }
```
Comment (hcoyote) [inline examples/private-vpc-no-zone/main.tf]: Should this and the promethus SG rule be conditional on state of enable_monitoring?

File: examples/private-vpc-no-zone/main.tf [hunk 29]
```diff
@@ -0,0 +1,299 @@
...
+      from_port       = -1
+      to_port         = -1
+      protocol        = "icmp"
+      enabled         = true
+      cidr_blocks     = ["0.0.0.0/0"]
+      self            = null
+      security_groups = [aws_security_group.client_sec_group.id]
+    }
+  }
+  ingress_rules = {
```
Comment (hcoyote) [inline examples/private-vpc-no-zone/main.tf:137]: Didn't notice it before, but I don't see the pandaproxy port listed

Comment (hcoyote) [reply examples/private-vpc-no-zone/main.tf:137]: (more specifically, we should probably have it added for the other examples as well as a common rule)

File: examples/private/main.tf (hunk 11 above)
Comment (hcoyote) [reply examples/private/main.tf:12]: fair enough.

File: vars.tf [hunk 30]
```diff
@@ -10,8 +10,8 @@ variable "availability_zone" {
   type        = list(string)
 }
 
-variable "clients" {
-  description = "Number of client hosts"
+variable "client_count" {
+  description = "Number of clients to build per AZ. Recommended 1"
```
Comment (hcoyote) [inline vars.tf:14]: Recommend at least 1 client for the cluster or 1 per az per cluster?

**Review by hcoyote — APPROVED** [review]

File: client.tf (hunk 26 above)
Comment (gene-redpanda) [reply client.tf:6]:
```  subnet_id                   = try(lookup(local.merged_subnets["client"], var.availability_zone[count.index % length(var.availability_zone)]), "")
```

So I'll go thru each function:

* try: this is a try-catch. If the contained functions throw an error it replaces the erroneous output with "". This results in us using the default subnet on the default vpc in practice. If there is no default vpc/subnet, then it errors out again and ends processing.

* lookup: we assume that the _count provided is intended to be created in each AZ rather than spread between them (so broker_count=3 means 3 per az). Because we have a list of AZ we can do a modulo on the count with the length of the AZ list to pick the correct AZ from the list of AZ. We then use that to do a lookup on the merged_subnets for client, which allows us to correctly pick the subnets for that AZ.

File: examples/private-vpc-no-zone/README.md (hunk 27 above)
Comment (gene-redpanda) [reply examples/private-vpc-no-zone/README.md:6]: Yes

File: examples/private-vpc-no-zone/main.tf (hunk 28 above)
Comment (gene-redpanda) [reply examples/private-vpc-no-zone/main.tf]: Can't conditionally include defaults

File: examples/private-vpc-no-zone/main.tf (hunk 29 above)
Comment (gene-redpanda) [reply examples/private-vpc-no-zone/main.tf:137]: Which port is that?

File: vars.tf (hunk 30 above)
Comment (gene-redpanda) [reply vars.tf:14]: 1 will be created per AZ. This isn't ideal imo but I don't have a better solution just yet.

File: examples/private-vpc-no-zone/main.tf (hunk 29 above)
Comment (r-vasquez) [reply examples/private-vpc-no-zone/main.tf:137]: 8082
